### PR TITLE
Adding aws_acm_export_certificates

### PR DESCRIPTION
**Pull request by diodonfrost**
##### SUMMARY

PR to introduce aws_acm_export_certificates module.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

aws_acm_export_certificates 

##### ADDITIONAL INFORMATION

This module allow to export [private certificates](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/acm.html#ACM.Client.export_certificate) from aws certificates manager via the Ansible framework.

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            certificate_arn=dict(type='str', required=True),
+            passphrase=dict(type='str', required=True)
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: Recommend using AnsibleAWSModule to get all the latest features - can just use `module.client` to get a client, and no need for region checking

Also without it your `module.fail_json_aws` won't work.

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+)
+
+
+try:
+    from botocore.exceptions import ClientError, ParamValidationError
+except ImportError:
+    pass  # caught by imported AnsibleAWSModule
+
+
+def normalize_certificates(acm_certificates):
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: This function could probably just be replaced by `camel_dict_to_snake_dict(acm_certificates)`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    kwargs = dict((k, v) for k, v in params.items() if v is not None)
+
+    try:
+        response = connection.export_certificate(**kwargs)
+    except (ClientError, ParamValidationError) as e:
+        module.fail_json_aws(e)
+
+    acm_certificates = normalize_certificates(response)
+    module.exit_json(certificates=acm_certificates)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: This should be singular

**Review by willthames — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 1 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
-from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.aws.core import AnsibleAWSModule
 from ansible.module_utils.ec2 import (
     boto3_conn,
     ec2_argument_spec,
@@ -76,7 +76,7 @@ except ImportError:
     pass  # caught by imported AnsibleAWSModule

 def main():
@@ -113,24 +112,12 @@ def main():
         )
     )

-    module = AnsibleModule(argument_spec=argument_spec)
+    module = AnsibleAWSModule(argument_spec=argument_spec)
+    connection = module.client('acm')

     if not HAS_BOTO3:
         module.fail_json(msg='boto3 and botocore are required.')

-    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
-    if region:
-        connection = boto3_conn(
-            module,
-            conn_type='client',
-            resource='acm',
-            region=region,
-            endpoint=ec2_url,
-            **aws_connect_kwargs
-        )
-    else:
-        module.fail_json(msg="region must be specified")
-
     export_acm_certificates(connection, module)
````
I used AnsibleAWSModule and `module.client`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 2 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
-def normalize_certificates(acm_certificates):
+def camel_dict_to_snake_dict(acm_certificates):
```
Use `camel_dict_to_snake_dict`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 3 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
-    acm_certificates = normalize_certificates(response)
-    module.exit_json(certificates=acm_certificates)
+    module.exit_json(certificates=camel_dict_to_snake_dict(response))
```

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        dict(
+            certificate_arn=dict(type='str', required=True),
+            passphrase=dict(type='str', required=True)
+        )
+    )
+
+    module = AnsibleAWSModule(argument_spec=argument_spec)
+    connection = module.client('acm')
+
+    if not HAS_BOTO3:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: Not required with AnsibleAWSModule

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        module.fail_json_aws(e)
+
+    module.exit_json(certificates=camel_dict_to_snake_dict(response))
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            certificate_arn=dict(type='str', required=True),
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: `type='str'` is the default, so not needed here.

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    type: dict
+    sample:
+      certificate: XXXXXXXXXXXXXXXXXXXX
+      certificate_chain: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+      private_key: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+'''
+
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import (
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:70]: These imports aren't needed but you will need `camel_dict_to_snake_dict` from `ansible.module_utils.ec2`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+- name: obtain specific privates certificates
+  aws_acm_export_certificates:
+    certificate_arn: "arn:aws:acm:us-east-1:123456789012:certificate/12345678-1234-1234-1234-123456789012"
+    passphrase: secretpassword
+  register: acm_certificates
+'''
+
+RETURN = '''
+certificates:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: This is still only one certificate. The name of the module, and the return value should be changed.

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+)
+
+
+try:
+    from botocore.exceptions import ClientError, ParamValidationError
+except ImportError:
+    pass  # caught by imported AnsibleAWSModule
+
+
+def camel_dict_to_snake_dict(acm_certificates):
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: Sorry, use `camel_dict_to_snake_dict` from `ansible.module_utils.ec2`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        'CertificateArn': module.params.get('certificate_arn'),
+        'Passphrase': module.params.get('passphrase')
+    }
+
+    kwargs = dict((k, v) for k, v in params.items() if v is not None)
+
+    try:
+        response = connection.export_certificate(**kwargs)
+    except (ClientError, ParamValidationError) as e:
+        module.fail_json_aws(e)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: Please add a `msg` value here e.g. "Failed to export certificates" just to make it a little clearer

**Review by willthames — CHANGES_REQUESTED** [review]
Still a few minor tweaks but overall this is looking very good.

**Comment by willthames** [comment]
This will also require a test suite (see test/integration/targets for existing ones) - it doesn't have to do much, just enough that we can validate that the module actually works. You may need to use the `command` module with the aws cli to create a certificate as we don't seem to have a module for creating acm certificates.

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 4 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
-    if not HAS_BOTO3:
-        module.fail_json(msg='boto3 and botocore are required.')
-
```

Removing boto3 check

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 5 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
@@ -107,17 +97,14 @@ def main():
     argument_spec = ec2_argument_spec()
     argument_spec.update(
         dict(
-            certificate_arn=dict(type='str', required=True),
-            passphrase=dict(type='str', required=True)
+            certificate_arn=dict(required=True),
+            passphrase=dict(required=True)
         )
     )
```
Removing `type='str`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 6 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:70]:
```
-def camel_dict_to_snake_dict(acm_certificates):                       
-    certificate = acm_certificates.get('Certificate', None)           
-    certificate_chain = acm_certificates.get('CertificateChain', None)
-    private_key = acm_certificates.get('PrivateKey', None)            
-    return {                                                          
-        'certificate': certificate,                                   
-        'certificate_chain': certificate_chain,                       
-        'private_key': private_key                                    
-    }                                                                 
-                                                                      
-                                                                      
 def export_acm_certificates(connection, module):                      
     params = {                                                        
```
Removing `camel_dict_to_snake_dict` function

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 7 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
-certificates:
-    description: The list of certificates and private key returned by the AWS Certificate Manager
+export:
+    description: Export the certificate, the certificate chain, and the encrypted private key associated with the public key embedded in the certificate
     returned: always
     type: dict
     sample:
```
Change ansible return value.

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 8 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
     try:
         response = connection.export_certificate(**kwargs)
-    except (ClientError, ParamValidationError) as e:
-        module.fail_json_aws(e)
+    except (ClientError, ParamValidationError) as error:
+        module.fail_json(msg="Couldn't obtain private certificates",
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(error.response))

     module.exit_json(certificates=camel_dict_to_snake_dict(response))
````

Use `camel_dict_to_snake_dict` from `ansible.module_utils.ec2`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 9 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
     try:
         response = connection.export_certificate(**kwargs)
-    except (ClientError, ParamValidationError) as e:
-        module.fail_json_aws(e)
+    except (ClientError, ParamValidationError) as error:
+        module.fail_json(msg="Couldn't obtain private certificates",
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(error.response))

     module.exit_json(certificates=camel_dict_to_snake_dict(response))
```
Add msg value during exception

**Comment by diodonfrost** [comment]
AWS acm command [export-certificate](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/acm.html#ACM.Client.export_certificate) only work with [private certificates](https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-request-private.html)

Private certificates can only be requested by [AWS Certificate Manager Private Certificate Authority ](https://docs.aws.amazon.com/acm-pca/latest/userguide/PcaWelcome.html)

I can make a test suite but you need a ACM PCA for it to work.

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    kwargs = dict((k, v) for k, v in params.items() if v is not None)
+
+    try:
+        response = connection.export_certificate(**kwargs)
+    except (ClientError, ParamValidationError) as error:
+        module.fail_json(msg="Couldn't obtain private certificates",
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(error.response))
+
+    module.exit_json(certificates=camel_dict_to_snake_dict(response))
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: This doesn't match docs and should still be singular certificate

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def main():
+    argument_spec = ec2_argument_spec()
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: not needed with AnsibleAWSModule

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        'CertificateArn': module.params.get('certificate_arn'),
+        'Passphrase': module.params.get('passphrase')
+    }
+
+    kwargs = dict((k, v) for k, v in params.items() if v is not None)
+
+    try:
+        response = connection.export_certificate(**kwargs)
+    except (ClientError, ParamValidationError) as error:
+        module.fail_json(msg="Couldn't obtain private certificates",
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: Use module.fail_json_aws(error, msg="Couldn't obtain private certificates")

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    - botocore
+    - python >= 2.6
+'''
+
+EXAMPLES = '''
+# Note: These examples do not set authentication details, see the AWS Guide for details.
+# Exporting a private certificate and private key (more details: https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-export-private.html#export-cli)
+
+- name: obtain specific privates certificates
+  aws_acm_export_certificates:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:46]: Rename the module to be singular as it only exports one certificate

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 10 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: Might also be better to just `module.exit_json(**camel_dict_to_snake_dict(response))`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    certificate_arn: "arn:aws:acm:us-east-1:123456789012:certificate/12345678-1234-1234-1234-123456789012"
+    passphrase: secretpassword
+  register: acm_certificates
+'''
+
+RETURN = '''
+export:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: As suggested below, maybe make `certificate`, `certificate_chain` and `private_key` top level return values

**Review by willthames — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 10 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
     try:
         response = connection.export_certificate(**kwargs)
     except (ClientError, ParamValidationError) as error:
-        module.fail_json(msg="Couldn't obtain private certificates",
-                         exception=traceback.format_exc(),
-                         **camel_dict_to_snake_dict(error.response))
+        module.fail_json_aws(error, msg="Couldn't obtain private certificate")

-    module.exit_json(certificates=camel_dict_to_snake_dict(response))
+    module.exit_json(**camel_dict_to_snake_dict(response))
```

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 11 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
 def main():
-    argument_spec = ec2_argument_spec()
-    argument_spec.update(
-        dict(
-            certificate_arn=dict(required=True),
-            passphrase=dict(required=True)
-        )
+    argument_spec = dict(
+        certificate_arn=dict(required=True),
+        passphrase=dict(required=True)
     )
-
-    module = AnsibleAWSModule(argument_spec=argument_spec)
+
+    module = AnsibleAWSModule(argument_spec=argument_spec, supports_check_mode=False,)
     connection = module.client('acm')
```

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 13 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:46]:
```
DOCUMENTATION = '''
 ---
-module: aws_acm_export_certificates
+module: aws_acm_export_certificate
 short_description: Exports a private certificate issued by a private certificate authority (CA) for use anywhere
 description:
     - Exports a private certificate issued by a private certificate authority (CA) for use anywhere
@@ -42,22 +42,26 @@ EXAMPLES = '''
 # Note: These examples do not set authentication details, see the AWS Guide for details.
 # Exporting a private certificate and private key (more details: https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-export-private.html#export-cli)

-- name: obtain specific privates certificates
-  aws_acm_export_certificates:
+- name: obtain specific privates certificate
+  aws_acm_export_certificate:
     certificate_arn: "arn:aws:acm:us-east-1:123456789012:certificate/12345678-1234-1234-1234-123456789012"
     passphrase: secretpassword
-  register: acm_certificates
+  register: acm_certificate
```

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 14 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
RETURN = '''
-export:
-    description: Export the certificate, the certificate chain, and the encrypted private key associated with the public key embedded in the certificate
+certificate:
+    Description: The base64 PEM-encoded certificate
     returned: always
-    type: dict
-    sample:
-      certificate: XXXXXXXXXXXXXXXXXXXX
-      certificate_chain: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
-      private_key: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    type: string
+certificate_chain:
+    Description: The base64 PEM-encoded certificate chain. This does not include the certificate that you are exporting
+    returned: always
+    type: string
+private_key:
+    Description: The encrypted private key associated with the public key in the certificate. The key is output in PKCS #8 format and is base64 PEM-encoded
+    returned: always
+    type: string
```

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 10 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: refactor `module.exit_json`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 11 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: remove `argument_spec = ec2_argument_spec()`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 13 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py:46]: rename module to `aws_acm_export_certificate`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 14 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]: return `certificate`, `certificate_chain` and `private_key`

File: lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py (hunk 12 above)
Comment (diodonfrost) [reply lib/ansible/modules/cloud/amazon/aws_acm_export_certificates.py]:
```
     try:
         response = connection.export_certificate(**kwargs)
     except (ClientError, ParamValidationError) as error:
-        module.fail_json(msg="Couldn't obtain private certificates",
-                         exception=traceback.format_exc(),
-                         **camel_dict_to_snake_dict(error.response))
+        module.fail_json_aws(error, msg="Couldn't obtain private certificate")
```

Fix `module.fail_json_aws` output

**Pull request closed without merging**